suitcase/utils/conftest.py

```diff
@@ -0,0 +1,123 @@
+        yield from trigger_and_read(dets)
+
+    yield from _internal_plan(dets)
+
+
+@pytest.fixture(params=[['det'], ['direct_img'], ['det', 'direct_img']],
+                scope='function')
+def detector_list(hw, request):
+    dets = [getattr(hw, det_name) for det_name in request.param]
+    return dets
+
+
+@pytest.fixture(params=['event', 'bulk_events', 'event_page'],
+                scope='function')
+def event_type(request):
+    return  request.param
+
+
+@pytest.fixture()
+def events_data(RE, detector_list, event_type):
+    '''A fixture that returnsevent data for a number of test cases.
+
+    Returns a list of (name, doc) tuples for the plan passed in as an arg.
+
+    Parameters
+    ----------
+    RE : object
+        pytest fixture object imported from `bluesky.test.conftest`
+    detector_list : list
+        pytest fixture defined in `suitcase.utils.conftest` which returns a
+        list of detectors
+    event_type : list
+        pytest fixture defined in `suitcase.utils.conftest` which returns a
+        list of 'event_types'.
+    '''
+
+    def _events_data_func(plan):
+        '''Generates data to be used for testing of suitcase.*.export(..)
+        functions
+
+        Parameters
+        ----------
+        plan : the plan to use to generate the test data
+
+        Returns
+        -------
+        collector : list
+            A list of (name, doc) tuple pairs generated by the run engine.
+        '''
+
+        # define the output lists and an internal list.
+        collector = []
+        event_list = []
+
+        # define the collector function depending on the event_type
+        if event_type == 'event':
+            def collect(name, doc):
+                collector.append((name, doc))
+                if name == 'event':
+                    event_list.append(doc)
+        elif event_type == 'event_page':
+            def collect(name, doc):
+                if name == 'event':
+                    event_list.append(doc)
+                elif name == 'stop':
+                    collector.append(('event_page',
+                                      event_model.pack_event_page(
+                                          *event_list)))
+                    collector.append((name, doc))
+                else:
+                    collector.append((name, doc))
+        elif event_type == 'bulk_events':
+            def collect(name, doc):
+                if name == 'event':
+                    event_list.append(doc)
+                elif name == 'stop':
+                    collector.append(('bulk_events', {'primary': event_list}))
+                    collector.append((name, doc))
+                else:
+                    collector.append((name, doc))
+        else:
+            raise UnknownEventType('Unknown event_type kwarg passed to '
+                                   'suitcase.utils.events_data')
+
+        # collect the documents
+        RE.subscribe(collect)
```

Comment:
Now I know how it works. Thanks for explanation, @awalter-bnl . I didn't check that event_type is defined as fixture parameter.

Reply:
@licode as I explained, each of the 'collect' functions is called in turn by the paramterization of the test. therefore when we ask the test to run it will run all 3 different 'collect' variations ('event', 'bulk_events' and 'event_page').